EventGenerator/src/CaloCalibGun_module.cc

```diff
@@ -0,0 +1,252 @@
+#include "CLHEP/Random/RandExponential.h"
+#include "CLHEP/Random/RandPoissonQ.h"
+
+
+#include "fhiclcpp/types/Atom.h"
+#include "messagefacility/MessageLogger/MessageLogger.h"
+
+//ROOT Includes
+#include <TMath.h>
+
+const double piconst = CLHEP::pi;
+using CLHEP::degree;
+using namespace std;
+using namespace mu2e;
+
+namespace mu2e {
+
+  class CaloCalibGun : public art::EDProducer {
+  public:
+    struct Config {
+      using Name=fhicl::Name;
+      using Comment=fhicl::Comment;
+      fhicl::Atom<unsigned> mean{Name("mean"),1};
+      fhicl::Atom<double> energy{Name("PhotonEnergy"),6.13};//MeV
+      fhicl::Atom<double> cosmin{Name("cosmin"),-1.};
+      fhicl::Atom<double> cosmax{Name("cosmax"),1.};
+      fhicl::Atom<double> phimin{Name("phimin"),0.};
+      fhicl::Atom<double> phimax{Name("phimax"), CLHEP::twopi };
+      fhicl::Atom<double> tmin{Name("tmin"),0.};
+      fhicl::Atom<double> tmax{Name("tmax"),1694.};
+      fhicl::Atom<bool> doHistograms{Name("doHistograms"), true };
+    };
+
+    using Parameters= art::EDProducer::Table<Config>;
+    explicit CaloCalibGun(const Parameters& conf);
+
+    virtual void produce(art::Event& event) override;
+
+    //----------------------------------------------------------------
+  private:
+
+    unsigned _mean;
+    double _energy;
+    double _cosmin;
+    double _cosmax;
+    double _phimin;
+    double _phimax;
+    double _tmin;
+    double _tmax;
+    
+    art::RandomNumberGenerator::base_engine_t& _engine;
+    CLHEP::RandFlat     _randFlat;
+    CLHEP::RandPoissonQ _randPoissonQ;
+    RandomUnitSphere    _randomUnitSphere;
+
+    const DiskCalorimeter *_cal;
+    //int                    _nPipes;
+    double                 _pipeRadius;
+    std::vector<double>    _pipeTorRadius;
+    std::vector<double>    _randomRad;
+    CLHEP::Hep3Vector      _zPipeCenter;
+
+  };
+
+  CaloCalibGun::CaloCalibGun(const Parameters& conf)
+    : EDProducer{conf}
+    , _mean{conf().mean()}
+    , _energy{conf().energy()}
+    , _cosmin{conf().cosmin()}
+    , _cosmax{conf().cosmax()}
+    , _phimin{conf().phimin()}
+    , _phimax{conf().phimax()}
+    , _tmin{conf().tmin()}
+    , _tmax{conf().tmax()}
+    , _engine{createEngine(art::ServiceHandle<SeedService>()->getSeed())}
+    , _randFlat{_engine}
+    , _randPoissonQ{_engine, _mean*1.0}
+    , _randomUnitSphere{_engine, _cosmin, _cosmax, 0, CLHEP::twopi}
+  {
+    produces<mu2e::GenParticleCollection>();
+    produces<mu2e::PrimaryParticle>();
+    produces <MCTrajectoryCollection>();
+
+    if (_mean < 0) throw cet::exception("RANGE") << "CaloCalibGun.mean must be non-negative "<< '\n';
+  }
+
+  //================================================================
+  void CaloCalibGun::produce(art::Event& event) {
+    std::unique_ptr<GenParticleCollection> output(new GenParticleCollection);
+    PrimaryParticle primaryParticles;
+    MCTrajectoryCollection mctc;
+
+    _cal =  &*GeomHandle<mu2e::DiskCalorimeter>();
```

Comment:
Calorimeter const &cal = *(GeomHandle<Calorimeter>()); is how I usually define these things i.e. in the Event Display. Would this work? or something more stable?

Reply:
As best I can tell there is no reason for _cal to be a member datum.  The required scope and lifetime is just the produce function.  So make it a local variable in the produce function this will work:
```Calorimeter const& cal = *(GeomHandle<DiskCalorimeter>())```;


If you need to get a pointer, not a reference, this will work:
```Calorimeter const* cal = GeomHandle<DiskCalorimeter>().get();```